### PR TITLE
[LLVM] Add 'ExpandVariadicsPass' to LTO default pipeline

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
This pass expands variadic functions into non-variadic function calls
according to the target ABI. Currently, this is used as the lowering for
the NVPTX and AMDGPU targets.

This pass is currently only run late in the target's backend. However,
during LTO we want to run it before the inliner pass so that the
expanded functions can be inlined using standard heuristics. This pass
is a no-op for unsupported targets, so this won't apply to any code that
isn't already using it.

### CODE REVIEW

**Review by arsenm — COMMENTED** [review]
Can we get a phase ordering test to show the inlining effect

**Comment by jhuber6** [comment]
> Can we get a phase ordering test to show the inlining effect

Good point, done.

File: llvm/test/Transforms/PhaseOrdering/varargs.ll [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+define internal i32 @vararg(i32 %first, ...) {
+entry:
+  %vlist = alloca ptr, align 8, addrspace(5)
+  %vlist.ascast = addrspacecast ptr addrspace(5) %vlist to ptr
+  call void @llvm.lifetime.start.p5(i64 8, ptr addrspace(5) %vlist)
+  call void @llvm.va_start.p0(ptr %vlist.ascast)
+  %vlist.promoted = load ptr, ptr addrspace(5) %vlist, align 8
+  %argp.next = getelementptr inbounds i8, ptr %vlist.promoted, i64 4
+  store ptr %argp.next, ptr addrspace(5) %vlist, align 8
+  %0 = load i32, ptr %vlist.promoted, align 4
```
Comment (arsenm) [inline llvm/test/Transforms/PhaseOrdering/varargs.ll:28]: Use named values

File: llvm/test/Transforms/PhaseOrdering/varargs.ll [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+;
+entry:
+  %call = tail call i32 (i32, ...) @vararg(i32 poison, i32 noundef 1, i32 noundef 2, i32 noundef 3)
+  ret i32 %call
+}
+
+define internal i32 @vararg(i32 %first, ...) {
+entry:
+  %vlist = alloca ptr, align 8, addrspace(5)
+  %vlist.ascast = addrspacecast ptr addrspace(5) %vlist to ptr
```
Comment (arsenm) [inline llvm/test/Transforms/PhaseOrdering/varargs.ll:22]: We shouldn't have these addrspacecasts. if clang is still emitting these, that really should be fixed

Comment (jhuber6) [reply llvm/test/Transforms/PhaseOrdering/varargs.ll:22]: Just what came out of a test C program, I'll clean it up.

Comment (jhuber6) [reply llvm/test/Transforms/PhaseOrdering/varargs.ll:22]: https://godbolt.org/z/88znf88bG was the source.

File: llvm/lib/Passes/PassBuilderPipelines.cpp [hunk 3]
```diff
@@ -1874,6 +1875,9 @@ PassBuilder::buildLTODefaultPipeline(OptimizationLevel Level,
   MPM.addPass(createModuleToFunctionPassAdaptor(std::move(PeepholeFPM),
                                                 PTO.EagerlyInvalidateAnalyses));
 
+  // Lower variadic functions for supported targets prior to inlining.
+  MPM.addPass(ExpandVariadicsPass(ExpandVariadicsMode::Optimize));
```
Comment (JonChesterfield) [inline llvm/lib/Passes/PassBuilderPipelines.cpp:1879]: The optimise version can run however often we like, on any subsets of the IR module. All good.

**Review by JonChesterfield — APPROVED** [review]

**Comment by JonChesterfield** [comment]
The original idea was to run the optimise per-tu instead of on the whole module at lto time, but until there's a non-gpu target using it in production the distinction doesn't matter too much. It's presumably at lto time here in order to combine application code with libc which seems legitimate

**Pull request merged**